lib/secp256k1/math.ex

```diff
@@ -0,0 +1,266 @@
+defmodule Bitcoinex.Secp256k1.Math do
+  @moduledoc """
+  Contains math utilities when dealing with secp256k1 curve points and scalars.
+
+  All of the addition and multiplication uses the secp256k1 curve paramaters.
+
+  Several of the jacobian multiplication and addition functions are borrowed heavily from https://github.com/starkbank/ecdsa-elixir/.
+  """
+  alias Bitcoinex.Secp256k1.{Params, Point}
+
+  @doc """
+  ipow performs integer pow,
+  where x is raised to the power of y.
+  """
+  @spec ipow(integer, integer) :: integer
+  def ipow(x, y, acc \\ 1)
+  def ipow(x, y, acc) when y > 0, do: ipow(x, y - 1, x * acc)
+  def ipow(_x, _y, acc), do: acc
+
+  @doc """
+  Inv performs the Extended Euclidean Algorithm to to find
+  the inverse of a number x mod n.
+  """
+  @spec inv(integer, integer) :: integer
+  def inv(x, _n) when x == 0, do: 0
+  def inv(x, n), do: inv(1, 0, modulo(x, n), n) |> modulo(n)
+
+  defp inv(lm, hm, low, high) when low > 1 do
+    r = div(high, low)
+
+    inv(
+      hm - lm * r,
+      lm,
+      high - low * r,
+      low
+    )
+  end
+
+  defp inv(lm, _hm, _low, _high) do
+    lm
+  end
+
+  @spec modulo(integer, integer) :: integer
+  def modulo(x, n) do
+    r = rem(x, n)
+    if r < 0, do: r + n, else: r
+  end
+
+  @doc """
+  multiply accepts a point P and scalar n and,
+  does jacobian multiplication to return resulting point.
+  """
+  def multiply(p, n) do
+    p
+    |> toJacobian()
+    |> jacobianMultiply(n)
+    |> fromJacobian()
+  end
+
+  @doc """
+  add accepts points p and q and,
+  does jacobian addition to return resulting point.
+  """
+  def add(p, q) do
```

Comment:
for the public functions like this it could make sense to use pattern matching in the function parameters to help ensure a certain amount of type safety

Reply:
Added guards for public math utilities and the Point module.